File: zerosum/zerosum.erl
%% zerosum.erl
%% Find lists that have sub-sets that sum to zero
%% 
%% Build a function that given a list of Integers in the range [-65000,65000], the function
%% 
%% returns true if any subset of the list summed is equal to zero. False otherwise.
%% 
%% Example:
%% 
%%  [0, 1, 2, -3] returns true. As 1+2+(-3)==0.
%%  [1, 2, 3, -8] returns false. As no subset summed is equal 0.
%%  [1, 4, 5, 2, -3] returns true.

-module(zerosum).
-export([subsets/1, zerosum/1]).

%% For a given list, find all sub-lists (i.e. combinations of elements)
subsets([]) ->
    [];
subsets(List) ->
    [Head | Tail] = List,
    TailSubSets = subsets(Tail),
    [[Head]] ++ TailSubSets ++ [[Head | Ts] || Ts <- TailSubSets].

zerosum(List) ->
    lists:any(
        fun(Subset) -> 
            lists:sum(Subset) == 0 end,
        subsets(List)
    ).

File: zerosum/zerosum.py
"""
Find subsets of a list that sum to zero
"""
from itertools import combinations

def subsets(numbers):
    """ Return a generator for all the subsets of a given list of numbers"""
    for i in xrange(1, len(numbers)):
        for combo in combinations(numbers, i):
            yield combo

def sumzero(numbers):
    """ Returns True if a subset sums to zero, False otherwise """
    for subset in subsets(numbers):
        if sum(subset) == 0:
            return True
    return False

def main():
    tests = (
        [0, 1, 2, -3],
        [1, 2, 3, -8],
        [1, 4, 5, 2, -3]
    )
    for testset in tests:
        if sumzero(testset):
            print "The set ", testset, " has a subset that sums to zero"
        else:
            print "The set ", testset, " does NOT have a subset that sums to zero"
            
if __name__ == "__main__":
    main()

